src/flag_gems/ops/repeat_interleave.py

```diff
@@ -53,3 +57,41 @@ def repeat_interleave_self_int(inp, repeats, dim=None, *, output_size=None):
     ndim = len(out_view_shape)
     copy_func.instantiate(ndim)(in_view, out0=out_view)
     return output
+
+
+@triton.jit
+def repeat_interleave_tensor_kernel(inputs_ptr, repeats_ptr, cumsum_ptr, out_ptr, size):
+    pid = tl.program_id(0)
+    mask = pid < size
+    cumsum = tl.load(cumsum_ptr + pid, mask, other=0)
+    inputs = tl.load(inputs_ptr + pid + 1, mask)
+    repeats = tl.load(repeats_ptr + pid + 1, mask)
+
+    tl.device_assert(repeats >= 0, "repeats can not be negative")
+
+    out_offsets = cumsum
+    for _ in range(repeats):
+        tl.store(out_ptr + out_offsets, inputs)
+        out_offsets += 1
+
+
+def repeat_interleave_tensor(repeats, *, output_size=None):
+    logging.debug("GEMS REPEAT_INTERLEAVE_TENSOR")
+
+    assert repeats.ndim == 1, "repeat_interleave only accept 1D vector as repeat"
+
+    cumsum = repeats.cumsum(axis=0)
+    result_size = cumsum[-1].item()
+
+    assert cumsum[0].item() >= 0, "repeats can not be negative"
+    assert result_size >= 0, "repeats can not be negative"
+
+    out = torch.zeros((result_size,), dtype=repeats.dtype, device=repeats.device)
+    size = repeats.size(0)
+    inputs = torch.arange(0, size, dtype=repeats.dtype, device=repeats.device)
+
+    grid = (size - 1,)
```

Comment:
This task partitioning along `repeats.size`(also the input's size at the repeat axis) may be in efficient when the repeat number is inbalanced.

I think we can consider a better task partitioning along the output's size at the repeat axis.

For example:

```text
Input: [a b c d]
repeats: [2 2 4 5]

Output: [a a b b c c c c d d d d d]
```

Instead of considering how many times and where to copy the input, we can consider the inverse, for each output, where to copy from the input.

```text
Input: [a b c d]
repeats: [2 2 4 5]

Output: [a a b b c c c c d d d d d]
Input Indices: [0 0 1 1 2 2 2 2 3 3 3 3 3]
```

We can generate the indices by comparing output indices with the cumsum of repeats and summing them.

|     |     |     |     |      |     |
| --- | --- | --- | --- | ---- | --- |
|     | >=2 | >=4 | 8>= | 13>= | sum |
| 0   | 0   | 0   | 0   | 0    | 0   |
| 1   | 0   | 0   | 0   | 0    | 0   |
| 2   | 1   | 0   | 0   | 0    | 1   |
| 3   | 1   | 0   | 0   | 0    | 1   |
| 4   | 1   | 1   | 0   | 0    | 2   |
| 5   | 1   | 1   | 0   | 0    | 2   |
| 6   | 1   | 1   | 0   | 0    | 2   |
| 7   | 1   | 1   | 0   | 0    | 2   |
| 8   | 1   | 1   | 1   | 0    | 3   |
| 9   | 1   | 1   | 1   | 0    | 3   |
| 10  | 1   | 1   | 1   | 0    | 3   |
| 11  | 1   | 1   | 1   | 0    | 3   |
| 12  | 1   | 1   | 1   | 0    | 3   |

With this input indices, we can easily parallelize the copying.

Reply:
Oh I see, repeat_interleave.Tensor returns that indices only without copying the input `[a,b,c,d]`.

Reply:
Tested when `repeats = torch.randint(0, 30, (1024, ), dtype=torch.int32, device="cuda")`, I've got the following results:
torch: 0.0993 ms
pr: 0.1781 ms
compare and reduce: 0.3164

Reply:
Oh, it is not as good as expected.